src/main/java/org/dataloader/DataLoaderHelper.java

```diff
@@ -546,4 +615,188 @@ private CompletableFuture<List<V>> setToValueCache(List<V> assembledValues, List
     private static <T> DispatchResult<T> emptyDispatchResult() {
         return (DispatchResult<T>) EMPTY_DISPATCH_RESULT;
     }
+
+    private class BatchObserverImpl implements BatchObserver<V> {
+        private final CompletableFuture<List<V>> valuesFuture;
+        private final List<K> keys;
+        private final List<Object> callContexts;
+        private final List<CompletableFuture<V>> queuedFutures;
+
+        private final List<K> clearCacheKeys = new ArrayList<>();
+        private final List<V> completedValues = new ArrayList<>();
+        private int idx = 0;
+        private boolean onErrorCalled = false;
+        private boolean onCompletedCalled = false;
+
+        private BatchObserverImpl(
+            CompletableFuture<List<V>> valuesFuture,
+            List<K> keys,
+            List<Object> callContexts,
+            List<CompletableFuture<V>> queuedFutures
+        ) {
+            this.valuesFuture = valuesFuture;
+            this.keys = keys;
+            this.callContexts = callContexts;
+            this.queuedFutures = queuedFutures;
+        }
+
+        @Override
+        public void onNext(V value) {
+            assert !onErrorCalled && !onCompletedCalled;
+
+            K key = keys.get(idx);
```

Comment:
So the assumption here is that the `Publisher` of events does so in key order - which makes sense - the other BatchLoader requires it to be in key order -  but this would need to be documented STRONGLY.

Reply:
I don't think we can know - its part of the contract - given the `List<K>` of keys each value returned will be matched with index in the order they arrive.  I dont think we can have guard rails beyond strong documentation.